### ISSUE TITLE
Utilize dreditor.org server for parsing patches

### ISSUE THREAD

**Issue by markhalliwell** [unicorn-fail/dreditor#161]
https://github.com/dreditor/dreditor.org/blob/7.x/sites/all/modules/custom/dreditor_review/dreditor_review.module

ref: #5, #30

**Comment by markhalliwell** [comment]
Fixed the dreditor_review so it works  again: https://github.com/dreditor/dreditor.org/commit/73f7a3a468d1c55a969a8bbec7e41936ded8d4eb

A `POST` request must be made, so for easier sharing, here are the results of that request:
https://www.runscope.com/public/c83c10b7-8cee-408e-9638-a43948590b42/05c3be9e-0a4d-4372-bdcd-03db3860fbfd

You can see the first patch is huge (https://drupal.org/node/2014017#comment-7663969, 2MB). After https://dreditor.org parses it and adds the necessary review markup to the patch it's around 6.6 MB. 

That's massive, but it's essential in understanding how much markup is actually being processed in JS.  Now, the beauty of this is that we can also send it from https://cdn.dreditor.org all gzipped up nicely. So the end result is still only transferring around 789KB, not that almost 7 MB file. 

Dreditor currently goes line by line and manipulates the browser's DOM. This is vastly expensive. So when the server parses the file, we can then do a single `$.replace()` and plop in the already parsed patch. No need to go line by line, on a user's browser.

What this also means is that we only have to parse a patch once. No need for each user to parse a patch when they click review.

# Marked up patch (cdn.dreditor.org)

https://cdn.dreditor.org/sites/default/files/review/2014017-14_0.patch.html
**Gzip compressed**

| Size | Time |
| --- | --- |
| 789 KB | 825 ms |

**Uncompressed**

| Size | Time* |
| --- | --- |
| 6.6 MB | 159 ms |

\* How long it took the browser to uncompress the gzipped contents of the file

# Original patch (drupal.org):

https://drupal.org/files/2014017-14_0.patch

**Gzip compressed**

| Size | Time |
| --- | --- |
| 377 KB | 1.34 s |

**Uncompressed**

| Size | Time* |
| --- | --- |
| 2.0 MB | 405 ms |

\* How long it took the browser to uncompress the gzipped contents of the file

**Comment by sun** [comment]
> 5.96 secs [Response time]

I assume that's the benchmark for comparing against the JS-only client-side code.

FWIW, the URLs definitely need to contain the version number of the Dreditor client, because the diff markup may be changed in the future.  If the markup changes on the server-side (+ latest client-side), then older Dreditor versions would break.  Likewise, without a version, the CDN cache would have to be invalidated.

I'm not sure I understand why this needs a separate Ajax/POST request.  Technically, this concept is nothing else than a HTTP Proxy that runs the content of the proxied resource (URI) through a filter.

That concept is so standard, it is even supported natively by web servers.  E.g., in Apache, that's the concept of output filters (as well as optional, built-in proxy modules).  I'm confident that nginx provides similar modules out of the box.

In other words:

```
                                             https://drupal.org/files/issues/some.patch
                                                     ↓
       https://dreditor.org/proxy/diff/1.2.7/https%3A%2F%2Fdrupal.org%2Ffiles%2Fissues%2Fsome.patch
                                                     ↓
                                     no     ←   file exists?   →   yes
                                     ↓                              ↓
                    download + filter/process + write               ↓
                                     ↓                              ↓
/var/www/sites/dreditor.org/proxy/diff/1.2.7/https%3A%2F%2Fdrupal.org%2Ffiles%2Fissues%2Fsome.patch
```

In essence, think of the image style architecture in Drupal.

**Comment by markhalliwell** [comment]
> In essence, think of the image style architecture in Drupal.

That is basically what `dreditor_review` does (and where I got the inspiration): https://github.com/dreditor/dreditor.org/blob/7.x/sites/all/modules/custom/dreditor_review/dreditor_review.module#L108-L109

That being said, I _did_ write that "module" in less than a couple hours one night a few months ago as really just a "proof of concept".... it most definitely could use a proper rewrite :)

> I'm not sure I understand why this needs a separate Ajax/POST request.

Client side perception/waiting. Instead of just waiting for the file to return (still 1-5 seconds when the file doesn't exist and it's actually parsing it), we can keep the client-side DOM from appearing to "lock" and slap a "Loading..." label while it _is_ loading. Pretty standard front-end implementation when dealing with varying sizes of content (especially when parsing original content is involved).

> FWIW, the URLs definitely need to contain the version number of the Dreditor client, because the diff markup may be changed in the future.

Agreed. I did not think about that aspect. That can easily be added as a parameter to be sent along with the initial request.

From our conversation (IRC) about PHP diff parsing, it's clear that each "version" that actually makes any changes would have to essentially have to extend the parser.

So the request may send version `1.3.5`, but because there have been no actual markup changes since version `1.2.7`, it would just use the latest available version: `class DreditorDiffParser_1_2_7 extends DreditorDiffParser` (exact naming conventions here are up in the air, but you get the idea).

Thoughts?

**Comment by markhalliwell** [comment]
In order to decrease latency (for smaller patches) we really should keep the JS parsing inside Dreditor. This then begs the question if we can just use Dreditor to always parse the patches.

For larger files, we should still send the request to https//dreditor.org; also sending the version (aka git ref) along with the request. We can simply checkout that specific version (if it doesn't already exist) of Dreditor and have it parse it on the server.

Granted, we'll likely have to perhaps build a separate app (NodeJS) for the server that could include [patch.review.js](https://github.com/dreditor/dreditor/blob/1.x/src/js/plugins/patch.review.js) of the checked out repo, but that's really just logistics I think.

I think this approach would help many of the above issues:
1. Keeping the parsing code in one place (consistency of output between server and local)
2. Keeping the parsing code in the same language (JS)
3. Ensuring the correct version of the parser is always used.

**Comment by sun** [comment]
> build a separate app (NodeJS) for the server that could include patch.review.js

That. + HTTP Proxy. = Sensible Architecture. :8ball:

**Comment by markhalliwell** [comment]
Dreditor is being decommissioned due to the lack of availability of current maintainers and the fact that Drupal.org is becoming more feature rich at a much faster rate than Dreditor itself now.

This decision was also partially made due to the amount of work (complete rewrite) it would take to create an "official" Firefox extension (#256).

If you already have the extension installed, it should continue working for a while until, inevitably,  the markup on Drupal.org changes. It will no longer be updated.

We understand that many of you love this extension, as do we. This isn't the end, in fact, it's a much brighter beginning. All the things we love about Dreditor can, and should, be moved into Drupal.org natively.

For more information, please follow: https://www.drupal.org/node/2779729

**Comment by subhojit777** [comment]
This issue was moved to dreditor/dreditor#25
